sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala

```diff
@@ -522,6 +539,119 @@ case class GpuMin(child: Expression) extends GpuAggregateFunction
   }
 }
 
+/** Min aggregation without `Nan` handling */
+case class GpuBasicMin(child: Expression) extends GpuMin(child)
+
+/** GpuMin for FloatType and DoubleType to handle `Nan`s.
+ *
+ * In Spark, `Nan` is the max float value, however in cuDF, the calculation
+ * involving `Nan` is undefined.
+ * We design a workaround method here to match the Spark's behaviour.
+ * The high level idea is:
+ *   if the column contains only `Nan`s or `null`s
+ *   then
+       if the column contains `Nan`
+ *     then return `Nan`
+ *     else return null
+ *   else
+ *     replace all `Nan`s with nulls;
+ *     use cuDF kernel to find the min value
+ */
+case class GpuFloatMin(child: Expression) extends GpuMin(child)
+  with GpuReplaceWindowFunction {
+  
+  override val dataType: DataType = child.dataType match {
+    case FloatType | DoubleType => child.dataType
+    case t => throw new IllegalStateException(s"child type $t is not FloatType or DoubleType")
+  }
+
+  protected val nan: Any = child.dataType match {
+    case FloatType => Float.NaN
+    case DoubleType => Double.NaN
+    case t => throw new IllegalStateException(s"child type $t is not FloatType or DoubleType")
+  }
+
+  protected lazy val updateAllNansOrNulls = CudfAll()
+  protected lazy val updateHasNan = CudfAny()
+  protected lazy val updateMinVal = new CudfMin(dataType)
+
+  protected lazy val mergeAllNansOrNulls = CudfAll()
+  protected lazy val mergeHasNan = CudfAny()
+  protected lazy val mergeMinVal = new CudfMin(dataType)
+
+  // Project 3 columns:
+  // 1. A boolean column indicating whether the values in `child` are `Nan`s or `null`s
+  // 2. A boolean column indicating whether the values in `child` are `Nan`s
+  // 3. Replace all `Nan`s in the `child` with `null`s
+  override lazy val inputProjection: Seq[Expression] = Seq(
+    GpuOr(GpuIsNan(child), GpuIsNull(child)),
+    GpuIsNan(child),
+    GpuNansToNulls(child)
+  )
+  // 1. Check if all values in the `child` are `Nan`s or `null`s
+  // 2. Check if `child` contains `Nan`
+  // 3. Calculate the min value on `child` with all `Nan`s has been replaced.
+  override lazy val updateAggregates: Seq[CudfAggregate] =
+    Seq(updateAllNansOrNulls, updateHasNan, updateMinVal)
+
+  // If the column only contains `Nan`s or `null`s
+  // Then
+  //   if the column contains `Nan`
+  //   then return `Nan`
+  //   else return `null`
+  // Else return the min value
+  override lazy val postUpdate: Seq[Expression] = Seq(
+    GpuIf(
+      updateAllNansOrNulls.attr, 
+      GpuIf(
+        updateHasNan.attr, GpuLiteral(nan, dataType), GpuLiteral(null, dataType)
+      ),
+      updateMinVal.attr
+    )
+  )
+
+  // Same logic as the `inputProjection` stage.
+  override lazy val preMerge: Seq[Expression] = Seq (
+    GpuOr(GpuIsNan(evaluateExpression), GpuIsNull(evaluateExpression)),
+    GpuIsNan(evaluateExpression),
+    GpuNansToNulls(evaluateExpression)
```

Comment:
Can we have something to explain why GpuNansToNulls is needed here?

Reply:
updated